Added version printing in neurodamus

iomaganaris opened the pull request:
- This PR goes in pair with https://bbpcode.epfl.ch/code/#/c/45205/
- Filter init.hoc to replace the UNKNOWN variables with the correct
  values
- Added clean=False to all versions of Neurodamus to keep the .git
  folder in the archive made by spack
- Added method get_commit_hash() in fetch_strategy.py to return the
  commit hash
- Added method get_commit_hash() in neurodamus-core and
  neurodamus-model to get the commit hash of those packages

iomaganaris committed:
Added version printing in neurodamus

- This PR goes in pair with https://bbpcode.epfl.ch/code/#/c/45205/
- Filter init.hoc to replace the UNKNOWN variables with the correct
  values
- Added clean=False to all versions of Neurodamus to keep the .git
  folder in the archive made by spack
- Added method get_commit_hash() in fetch_strategy.py to return the
  commit hash
- Added method get_commit_hash() in neurodamus-core and
  neurodamus-model to get the commit hash of those packages

ferdonline commented on var/spack/repos/builtin/packages/neurodamus-core/package.py:
```diff
@@ -36,6 +36,11 @@ class NeurodamusCore(Package):
     depends_on('py-enum34',        type=('run',), when='^python@2.4:2.7.999,3.1:3.3.999')
     depends_on('py-lazy-property', type=('run'), when='+python')
 
+    def get_commit_hash(self):
```
Maybe no need to implement a function that is only called once and can be in a single line
`git_commit_hash = self.fetcher[0].get_commit_hash()`

iomaganaris replied:
I think you are right, since they are used only once

ferdonline commented on var/spack/repos/builtin/packages/neurodamus-hippocampus/package.py:
```diff
@@ -11,6 +11,6 @@ class NeurodamusHippocampus(NeurodamusModel):
-    version('develop', git=git, branch='master', submodules=True)
-    version('0.2', git=git, tag='0.2', submodules=True, preferred=True)
-    version('0.1', git=git, tag='0.1', submodules=True)
+    version('develop', git=git, branch='master', submodules=True, clean=False)
```
Would be nice to set `clean=False` once in neurodamus-model (overriding version) so that all inherit this setting

iomaganaris replied:
After conversation we decided it is not easily done due to the definition of Version and the DirectiveMeta which are used in spack. Could be done in the future.

iomaganaris committed:
Removed get_commit_hash() from neurodamus-core and neurodamus-model
recipes and replaced its calls with a single line function call

pramodk commented on var/spack/repos/builtin/packages/neurodamus-core/package.py:
```diff
@@ -45,6 +46,10 @@ def install(self, spec, prefix):
 
-    def setup_environment(self, spack_env, run_env):
+        filter_file(r'UNKNOWN_CORE_VERSION', r'%s' % spec.version, prefix.hoc.join('defvar.hoc'))
+        git_commit_hash = self.fetcher[0].get_commit_hash()
```
I think we should add small check if `fetcher` is `git fetcher`. We know that it's git package but good to have a check. (if you want, you can move fetcher related line in senate method).

iomaganaris replied:
You are right. It's now done :+1:

iomaganaris committed:
Added method that checks if the fetcher used to get the git_commit
is an object of GitFetchStrategy and has the same url as the package

iomaganaris committed:
Changed preferred version of neurodamus-core to 2.3.4

- This tag encapsulates the addition of NEURODAMUS_MODEL,
  NEURODAMUS_MODEL_VERSION, NEURODAMUS_MODEL_HASH,
  NEURODAMUS_CORE_VERSION and NEURODAMUS_CORE_HASH variables in
  defvar.hoc that are set by neurodamus-model spack recipe
- Addition of get_version() function in defvar.hoc, which is called
  in init.hoc to print the versions of neurodamus and neurodamus-core
  in the beggining of the simulation

Sample output: Version: neurodamus-neocortex@0.1 ['c40093b83bec104dc8a4966fb3632a856fde2f58'] Core: 2.3.4 ['c40aef90f74c3d47e23b54d6612cf7961417d38b']

pramodk commented:
@matz-e : could you take a look at this?

matz-e commented on var/spack/repos/builtin/packages/neurodamus-core/package.py:
```diff
@@ -3,17 +3,18 @@
-    version('2.3.3', git=git, tag='2.3.3', preferred=True)
-    version('2.2.1', git=git, tag='2.2.1')
+    version('develop', git=git, branch='master', clean=False)
+    version('2.3.4', git=git, tag='2.3.4', clean=False, preferred=True)
```
The newest version is always preferred
```suggestion
    version('2.3.4', git=git, tag='2.3.4', clean=False)
```

matz-e commented on var/spack/repos/builtin/packages/neurodamus-hippocampus/package.py:
```diff
@@ -11,6 +11,6 @@ class NeurodamusHippocampus(NeurodamusModel):
-    version('0.2', git=git, tag='0.2', submodules=True, preferred=True)
-    version('0.1', git=git, tag='0.1', submodules=True)
+    version('develop', git=git, branch='master', submodules=True, clean=False)
+    version('0.2', git=git, tag='0.2', submodules=True, preferred=True, clean=False)
```
Same as above?
```suggestion
    version('0.2', git=git, tag='0.2', submodules=True, clean=False)
```

matz-e commented on var/spack/repos/builtin/packages/neurodamus-mousify/package.py:
```diff
@@ -11,5 +11,5 @@ class NeurodamusMousify(NeurodamusModel):
-    version('develop', git=git, branch='master', submodules=True)
-    version('0.1', git=git, tag='0.1', submodules=True, preferred=True)
+    version('develop', git=git, branch='master', submodules=True, clean=False)
+    version('0.1', git=git, tag='0.1', submodules=True, preferred=True, clean=False)
```
```suggestion
    version('0.1', git=git, tag='0.1', submodules=True, clean=False)
```

matz-e commented on var/spack/repos/builtin/packages/neurodamus-neocortex/package.py:
```diff
@@ -11,8 +11,8 @@ class NeurodamusNeocortex(NeurodamusModel):
-    version('develop', git=git, branch='master', submodules=True)
-    version('0.1', git=git, tag='0.1', submodules=True, preferred=True)
+    version('develop', git=git, branch='master', submodules=True, clean=False)
+    version('0.1', git=git, tag='0.1', submodules=True, preferred=True, clean=False)
```
```suggestion
    version('0.1', git=git, tag='0.1', submodules=True, clean=False)
```

matz-e commented on var/spack/repos/builtin/packages/neurodamus-thalamus/package.py:
```diff
@@ -11,5 +11,5 @@ class NeurodamusThalamus(NeurodamusModel):
-    version('develop', git=git, branch='master', submodules=True)
-    version('0.1', git=git, tag='0.1', submodules=True, preferred=True)
+    version('develop', git=git, branch='master', submodules=True, clean=False)
+    version('0.1', git=git, tag='0.1', submodules=True, preferred=True, clean=False)
```
```suggestion
    version('0.1', git=git, tag='0.1', submodules=True, clean=False)
```

matz-e commented on lib/spack/spack/fetch_strategy.py:
```diff
@@ -712,6 +712,11 @@ def fetch(self):
 
+    def get_commit_hash(self):
+        git_exe = self.git.name
+        git_commit = subprocess.check_output([git_exe, 'rev-parse', 'HEAD']).split()
```
I have to look into this a bit more, but spack has this `command` stuff. No need for `subprocess`:

https://github.com/spack/spack/blob/4ecb465a366de3f722aeb1525b22c0d22aefceae/var/spack/repos/builtin/packages/spark/package.py#L51-L53

`check_output` is not available for earlier Python versions.

iomaganaris replied:
I was trying to do that but I failed in the past. Now it works fine.
Thanks for pointing that out :+1:

iomaganaris committed:
Addressed review comments

- Removed prefered=True from packages, as the latest version is
  selected by default
- Removed subprocess.check_output from get_commit_hash() in fetch_strategy
  and used self.git command instead to print to str

matz-e commented on lib/spack/spack/fetch_strategy.py:
```diff
@@ -40,7 +40,7 @@
 from spack.version import Version, ver
 from spack.util.compression import decompressor_for, extension
-
+import subprocess
```
```suggestion
```

iomaganaris replied:
You are right, I just forgot this :/

ferdonline commented on lib/spack/spack/fetch_strategy.py:
```diff
@@ -40,7 +40,7 @@
 from spack.version import Version, ver
 from spack.util.compression import decompressor_for, extension
-
+import subprocess
```
now we dont need this import

iomaganaris replied:
You are right.. I just forgot this :/

ferdonline commented on var/spack/repos/builtin/packages/neurodamus-core/package.py:
```diff
@@ -36,6 +37,9 @@ class NeurodamusCore(Package):
     depends_on('py-lazy-property', type=('run'), when='+python')
 
+    def get_git_fetcher(self):
+        return next((fetcher for fetcher in self.fetcher if (isinstance(fetcher, GitFetchStrategy) and fetcher.url == self.git)), None)
```
From https://github.com/BlueBrain/spack/blob/develop/lib/spack/spack/package.py#L665 I understand it is guaranteed the root fetcher to be the first one in the list. Maybe we could just check it's git. 
That also simplifies the line (a very nice 1-liner though!) :)
Something like this? 

```suggestion
        root_fetcher = self.fetcher[0]
        return root_fetcher if isinstance(root_fetcher, GitFetchStrategy) else None
```

matz-e replied:
I like that, a lot simpler!

iomaganaris replied:
I agree with you, I just added the other checks just to be sure

ferdonline commented on var/spack/repos/builtin/packages/neurodamus-model/package.py:
```diff
@@ -93,6 +94,9 @@ def build(self, spec, prefix):
                                                  loadflags=link_flag))
         os.chmod(_BUILD_NEURODAMUS_FNAME, 0o770)
 
+    def get_git_fetcher(self):
```
same as above

ferdonline commented on var/spack/repos/builtin/packages/neurodamus-core/package.py:
```diff
@@ -3,17 +3,18 @@
-    version('2.3.3', git=git, tag='2.3.3', preferred=True)
-    version('2.2.1', git=git, tag='2.2.1')
+    version('develop', git=git, branch='master', clean=False)
+    version('2.3.4', git=git, tag='2.3.4', clean=False)
```
I'm not very found of the fact we have to leave dust behind...
When fetching we could retrieve the commit_hash and save it in a fetcher variable...

matz-e replied:
That's what we have to do when using `setuptools-scm`, too. Not very elegant, but sometimes necessary.

ferdonline replied:
Wait, this clean=False does anything else than avoiding deleting the stage dir? I meant, in GitFetchStrategy.fetch() add the line to retrieve the commit. Then the clean could happen normally (?)

matz-e replied:
The stage directory still gets deleted. This is for source caching, where `.git` is not included in the cached tarball. When using a source mirror, this is harmful.

ferdonline replied:
> This is for source caching, where .git is not included in the cached tarball.

Would it be too bad to change that behavior?

ferdonline replied:
btw, cached source tarballs dont get named after the commit hash?

iomaganaris replied:
> btw, cached source tarballs dont get named after the commit hash?

Unfortunately not. For example for `neurodamus-neocortex` there is this:
```
==> Using cached archive: /home/magkanar/spack/var/spack/cache/neurodamus-neocortex/neurodamus-neocortex-0.1.tar.gz
```

matz-e replied:
> > This is for source caching, where .git is not included in the cached tarball.
> 
> Would it be too bad to change that behavior?

Randomly deleting things from a source checkout and not expecting things to break seems bad behavior to me :)

With `clean=False`, the SCM directories will not be removed when creating the source tarball. The caching mechanism here is fairly generic, and I think it works fine to keep the SCM directory (to conserve space, it would be more elegant to have a shallow clone). Adding a commit hash to a filename seems to ask for different kinds of trouble.

ferdonline replied:
> The stage directory still gets deleted.

I hadn't understood that... So the case is less worrying.

> To conserve space, it would be more elegant to have a shallow clone

It seems they do it:
https://github.com/BlueBrain/spack/blob/develop/lib/spack/spack/fetch_strategy.py#L674

So if we really want to tackle that 

- we could eventually include a git metadata file in the cached archive (sounds excessive)
- find a way to set make clean=False to all versions. That was the original idea but it showed to be a bit more complicated than expected

matz-e replied:
>  * find a way to set make clean=False to all versions. That was the original idea but it showed to be a bit more complicated than expected

I'm fine with how it is currently. See also spack#9436.

ferdonline replied:
> See also [spack#9436](https://github.com/spack/spack/pull/9436).

Ah, I didn't know you introduced it! :)

Another thought. Since we are "hardcoding" the versions in source files, why not doing that before the archive phase?

iomaganaris replied:
This is a good idea. I was thinking if it is possible to use something like `@run_before('archive')` and do then the replacement of the values but I don't know when exactly does the `archive` takes place and if I can refer to it somehow. I am looking at it..

matz-e reviewed (commented):
Had some comments like @ferdonline that I apparently didn't submit…

iomaganaris committed:
Addressed some of the review comments

- Removed checks for all fetchers and return only the first one if
  it is a GitFetchStrategy fetcher in get_git_fetcher()
- Removed `import subprocess` in fetch_strategy.py since it's not
  needed

ferdonline reviewed (approved):
LGTM.
Attempt to get commit hashes without requiring clean=False can be worked in another PR

matz-e commented on lib/spack/spack/fetch_strategy.py:
```diff
@@ -712,6 +711,10 @@ def fetch(self):
         super(GitFetchStrategy, self).archive(destination, exclude='.git')
 
+    def get_commit_hash(self):
+        git_commit = self.git('rev-parse', 'HEAD', output=str)
```
Could be shorter:
```suggestion
        return self.git('rev-parse', 'HEAD', output=str).strip()
```

matz-e replied:
And delete the next line

matz-e replied:
Remove instead.

matz-e commented on var/spack/repos/builtin/packages/neurodamus-core/package.py:
```diff
@@ -3,17 +3,18 @@
 import os
 import shutil
-
+from spack.fetch_strategy import GitFetchStrategy
```
Maybe move to after line 40.

matz-e replied:
Actually just ditch it. See my later comment.

matz-e commented on var/spack/repos/builtin/packages/neurodamus-model/package.py:
```diff
@@ -4,6 +4,7 @@
 from spack.pkg.builtin.sim_model import SimModel
 import shutil
 import os
+from spack.fetch_strategy import GitFetchStrategy
```
Mabye move to after line 97.

matz-e commented on var/spack/repos/builtin/packages/neurodamus-model/package.py:
```diff
@@ -93,6 +94,10 @@ def build(self, spec, prefix):
                                                  loadflags=link_flag))
         os.chmod(_BUILD_NEURODAMUS_FNAME, 0o770)
 
+    def get_git_fetcher(self):
```
Actually I'd consider moving the two lines of this function into line 137, or factor them out.

matz-e commented on var/spack/repos/builtin/packages/neurodamus-core/package.py:
```diff
@@ -36,6 +37,10 @@ class NeurodamusCore(Package):
     depends_on('py-enum34',        type=('run',), when='^python@2.4:2.7.999,3.1:3.3.999')
     depends_on('py-lazy-property', type=('run'), when='+python')
 
+    def get_git_fetcher(self):
```
Same as below, maybe just move these two line to 54 (and replace that one)

matz-e commented on var/spack/repos/builtin/packages/neurodamus-core/package.py:
```diff
@@ -45,6 +50,12 @@ def install(self, spec, prefix):
-    def setup_environment(self, spack_env, run_env):
+        filter_file(r'UNKNOWN_CORE_VERSION', r'%s' % spec.version, prefix.hoc.join('defvar.hoc'))
+        git_fetcher = self.get_git_fetcher()
+        if git_fetcher is not None:
```
Instead of all these indirections, you could just ask for forgiveness:
```python
try:
    commit_hash = self.fetcher[0].get_commit()
    filter_file(r'UNKNOWN_CORE_HASH', r'\'{}\''.format(commit), prefix.hoc.join('defvar.hoc'))
except:
    pass
```
IMO that's a lot cleaner, and generic enough that the underlying RCS does not matter.

matz-e replied:
I think that should be done for both files. Do we care that it's a `Git<Whatever>Fetcher`? All we care about is whether or not it provides us with a `commit`, or even better, `version` thing. You could catch an `AttributeError` to be on the safer side and re-raise if it's not complaining about `get_commit`, but I think this short version conveys everything needed and would serve our purpose.

matz-e commented on lib/spack/spack/fetch_strategy.py:
```diff
@@ -712,6 +711,10 @@ def fetch(self):
     def archive(self, destination):
         super(GitFetchStrategy, self).archive(destination, exclude='.git')
 
+    def get_commit_hash(self):
```
```suggestion
    def get_commit(self):
```
No need to be specific here.

matz-e reviewed (changes requested):
Some minor final nitpicks. I think the code can be simplified a bit more. Makes it cleaner and easier to follow?

iomaganaris committed:
Addressed review comments about replacing the GitFetcher checking with exception

- Removed get_git_fetcher and replaced it with fetcher[0]
- The case of being git or not is checked by catching an exception
